src/fastgltf.cpp

```diff
@@ -1303,366 +1313,449 @@ fg::Expected<fg::Asset> fg::Parser::parse(simdjson::dom::object root, Category c
 
-	asset.accessors.reserve(accessors.size());
+	// asset.accessors.reserve(accessors.size());
     for (auto accessorValue : accessors) {
-        // Required fields: "componentType", "count"
-        Accessor accessor = {};
-        dom::object accessorObject;
+        ondemand::object accessorObject;
         if (accessorValue.get_object().get(accessorObject) != SUCCESS) {
             return Error::InvalidGltf;
         }
 
-        std::uint64_t componentType;
-        if (accessorObject["componentType"].get_uint64().get(componentType) != SUCCESS) {
-            return Error::InvalidGltf;
-        }
-		accessor.componentType = getComponentType(static_cast<std::underlying_type_t<ComponentType>>(componentType));
-        if (accessor.componentType == ComponentType::Double && !hasBit(options, Options::AllowDouble)) {
-            return Error::InvalidGltf;
-        }
-
-        std::string_view accessorType;
-        if (accessorObject["type"].get_string().get(accessorType) != SUCCESS) {
-            return Error::InvalidGltf;
-        }
-		accessor.type = getAccessorType(accessorType);
-
-        std::uint64_t accessorCount;
-        if (accessorObject["count"].get_uint64().get(accessorCount) != SUCCESS) {
-            return Error::InvalidGltf;
-        }
-		accessor.count = static_cast<std::size_t>(accessorCount);
+        auto& accessor = asset.accessors.emplace_back();
+        accessor.byteOffset = 0UL;
+        accessor.normalized = false;
+
+		// Function for parsing the min and max arrays for accessors
+        auto parseMinMax = [&](simdjson::ondemand::array& array, decltype(Accessor::max)& ref) -> fastgltf::Error {
+            decltype(Accessor::max) variant;
+
+            using double_vec = std::variant_alternative_t<1, decltype(Accessor::max)>;
+            using int64_vec = std::variant_alternative_t<2, decltype(Accessor::max)>;
+
+            // It's possible the min/max fields come before the accessor type is declared, in which case we don't know the size.
+			// This single line is actually incredibly important as this function without it takes up roughly 15% of the entire
+			// parsing process on average due to vector resizing.
+			// When no exact count is known (accessor type field comes after min/max fields) we'll just count them and take the penalty.
+            auto initialCount = accessor.type == AccessorType::Invalid ? array.count_elements().value() : getNumComponents(accessor.type);
+            if (accessor.componentType == ComponentType::Float || accessor.componentType == ComponentType::Double) {
+                auto vec = FASTGLTF_CONSTRUCT_PMR_RESOURCE(double_vec, resourceAllocator.get(), 0);
+				vec.reserve(initialCount);
+				variant = std::move(vec);
+            } else {
+                auto vec = FASTGLTF_CONSTRUCT_PMR_RESOURCE(int64_vec, resourceAllocator.get(), 0);
+				vec.reserve(initialCount);
+				variant = std::move(vec);
+            }
 
+			for (auto element : array) {
```

Comment:
I am not sure I understand this code, but I would expect that you could do...

```
if(accessor.componentType == ComponentType::Double) {
  auto v = std::get<double_vec>(variant);
  for (auto element : array) {
    // error handing as needed
    v.push_back(element.get_double());
  }
}
```

And so forth. The number type is a fallback for complicated scenarios, where you need to branch according to the number type, but if you know that a double is good enough in some instances, then just grab that.

Reply:
Okay, so a few things first:
- I should actually be down casting to single precision in all cases:
    > For floating-point components, JSON-stored minimum and maximum values represent single precision floats and SHOULD be rounded to single precision before usage to avoid any potential boundary mismatches.
- JSON does not make the distinction between integers and floating point numbers, as you know. Therefore, based on `accessor.componentType`, I have to cast either to float or to an integer type. The glTF spec specifically says this:
    > Array elements MUST be treated as having the same data type as accessor’s componentType.
    
    From that, I understand that if the component type specifies an integer but the value is `3.5` for whatever reason in the JSON, I have to just cast that to an integer. The specification makes no further clarifications on how the values should be treated as far as I can tell.
    
As I don't know how simdjson parses or identifies the numbers internally, would you say it'd be safe to just have this instead to avoid the `ondemand::number` object?
```cpp
switch (accessor.componentType) {
    case ComponentType::Double:
        v.emplace_back(element.get_double();
        break;
    case ComponentType::Float:
        ...
}
```

Reply:
Converting to float from double is typically done with a single instruction (e.g., cvtsd2ss). It is a high latency operation, but not overly expensive. So if you don't need to immediately check the result, it should be quite cheap. 

> would you say it'd be safe to just have this 

I wouldn't use a switch case if I could avoid it to handle individual elements. I would have distinct code paths based on `componentType`. It depends on the context, with repeated calls to a switch/case can be detrimental to the performance. Not always, it depends on how the compiler decides to handle it and how precisely you wrote the code, but to be safe, assuming that a switch/case adds significant overhead because it might. It is not the switch case per se that's the problem, it is that it prevents some useful optimizations by the compiler.

Of course, trying to parse `"3.5"` as an integer will cause an error, but you could handle this with an exceptional case (using a cold path if needed).

Reply:
> So the proposed code isn't a possibility

My proposal is that you try it and benchmark. I submit to you that you need to run experiments to find the right design, even if it means temporarily having code that does not solve the right problem. You need to identify the bottleneck and that's damn difficult if you are just starting at high-level code.

Reply:
I wrote a small little benchmark which tried to parse the same JSON array which contains integers and doubles (50-50 distribution). Turns out `get_number` is ever so slightly faster in my little experiment, even with the switch-case covering those three cases.